src/osdproc.c

```diff
@@ -692,79 +708,155 @@ void draw_time() {
-        if (++osd_alt_cnt >= 75) {
-          osd_home_alt = osd_alt;           // take this stable osd_alt as osd_home_alt
+  }    
+}
+
+void set_home_altitude_if_unset()
+{
+    if (osd_got_home == 1)
+    {
+        // JRChange: osd_home_alt: check for stable osd_alt (must be stable for 75*40ms = 3s)
+        // we can get the relative alt from mavlink directly.
+        if (osd_alt_cnt < 75) {
+          if (fabs(osd_alt_prev - osd_alt) > 0.5) {
+            osd_alt_cnt = 0;
+            osd_alt_prev = osd_alt;
+          } else {
+            osd_alt_cnt++;
+            if (osd_alt_cnt >= 75) {
+              osd_home_alt = osd_alt;           // take this stable osd_alt as osd_home_alt
+            }
+          }
         }
-      }
     }
+}
 
-
-    float rads = fabs(osd_home_lat) * 0.0174532925;
-    double scaleLongDown = cos(rads);
-    double scaleLongUp   = 1.0f/cos(rads);
-
-    dstlat = fabs(osd_home_lat - osd_lat) * 111319.5f;
-    dstlon = fabs(osd_home_lon - osd_lon) * 111319.5f * scaleLongDown;
-    dstsqrt = dstlat * dstlat + dstlon * dstlon;
-    osd_home_distance = sqrt(dstsqrt) / 10000000.0f;
-
-    dstlon = (osd_home_lon - osd_lon); //OffSet_X
-    dstlat = (osd_home_lat - osd_lat) * scaleLongUp; //OffSet Y
-
-    long bearing = 90 + (atan2(dstlat, -dstlon) * 57.295775); //absolut home direction
-    if(bearing < 0) bearing += 360;//normalization
-    bearing = bearing - 180;//absolut return direction
-    if(bearing < 0) bearing += 360;//normalization
-    osd_home_bearing = bearing;
-  }
-
-  //distance
-  if (eeprom_buffer.params.CWH_home_dist_en == 1 && shownAtPanel(eeprom_buffer.params.CWH_home_dist_panel)) {
-    float tmp = osd_home_distance * convert_distance;
-    if (tmp < convert_distance_divider)
-      sprintf(tmp_str, "H %d%s", (int)tmp, dist_unit_short);
-    else
-      sprintf(tmp_str, "H %0.2f%s", (double)(tmp / convert_distance_divider), dist_unit_long);
-
-    write_string(tmp_str, eeprom_buffer.params.CWH_home_dist_posX, eeprom_buffer.params.CWH_home_dist_posY, 0, 0, TEXT_VA_TOP, eeprom_buffer.params.CWH_home_dist_align, 0, SIZE_TO_FONT[eeprom_buffer.params.CWH_home_dist_fontsize]);
-  }
-  if ((wp_number != 0) && (eeprom_buffer.params.CWH_wp_dist_en) && shownAtPanel(eeprom_buffer.params.CWH_wp_dist_panel)) {
-    float tmp = wp_dist * convert_distance;
-    if (tmp < convert_distance_divider)
-      sprintf(tmp_str, "WP %d%s", (int)tmp, dist_unit_short);
-    else
-      sprintf(tmp_str, "WP %0.2f%s", (double)(tmp / convert_distance_divider), dist_unit_long);
-
-    write_string(tmp_str, eeprom_buffer.params.CWH_wp_dist_posX, eeprom_buffer.params.CWH_wp_dist_posY, 0, 0, TEXT_VA_TOP, eeprom_buffer.params.CWH_wp_dist_align, 0, SIZE_TO_FONT[eeprom_buffer.params.CWH_wp_dist_fontsize]);
+// After compartmentalizing it, I'm not sure why this is a discrete routine. 
+// Why bundle these things together? And what does CWH stand for anyhow? -- SLG
+void draw_CWH(void) {
+  
+  set_home_position_if_unset();
```

Comment:
I think we should do this in the loop before starting to draw panels on the screen.
That way you can use the info in any panel and the initialization of state could be moved to a central point (over time)